### PR TITLE
Ball in cup env

### PR DESCRIPTION
**Pull request by jjshoots**
To-do:
- [x] make observation space consistent
- [x] verify learning
- [x] fix colision bug

### CODE REVIEW

**Comment by jjshoots** [comment]
@defrag_bambino , perhaps an interesting point of discussion, would it make sense for the environment to only be considered successful if the UAV successfully balances the ball for a number of seconds or achieves (close to) 0 velocity on it?

**Comment by jjshoots** [comment]
I made the following changes:
1. Add ball relative velocity to drone observation space
2. Tweaked reward and termination structure
3. Added it to tests

**Comment by defrag-bambino** [comment]
> @defrag-bambino , perhaps an interesting point of discussion, would it make sense for the environment to only be considered successful if the UAV successfully balances the ball for a number of seconds or achieves (close to) 0 velocity on it?

That is certainly an option. The way I intented to use it was to have some sort of basket on top of the drone. Once it fell in, the ball would not be able to move in there anyways.
Maybe your idea could be the difficult version of the env? So we'll have (atleast) two versions.

**Comment by jjshoots** [comment]
Currently, the drone is still not able to learn the right behaviour using my learning stack and the current reward function. Lemme poke around a bit more to get a stable configuration.

**Comment by defrag-bambino** [comment]
> Currently, the drone is still not able to learn the right behaviour using my learning stack and the current reward function. Lemme poke around a bit more to get a stable configuration.

What is does it do instead? Just not move at all? Or swing the ball around wildly?
I had the following issues and solved them:
- Swinging around itself. The drone would move such that the ball kept swinging around and around itself. Solution: end the episode once the ball was above the drone and went back down.
- No swinging at all. Solution: add some random ball rotation at the beginning, since then the pendulum will already be in a moving state and the agent will quickly realize that this gives reward.

**Comment by jjshoots** [comment]
Basically anything in between not moving and swinging wildly. I am hesitant to implement the first fix yet because that adds an additional task to the environment. The second fix sounds reasonable, I'll give it a shot.

**Comment by defrag-bambino** [comment]
> Basically anything in between not moving and swinging wildly. I am hesitant to implement the first fix yet because that adds an additional task to the environment. The second fix sounds reasonable, I'll give it a shot.

Ok.
The first fix came expecially handy once it learned to exploit the reward structure by spinning the ball around itself. This is not only undesired in terms of the goal, but also unrealistic because the rope would get tangled once the ball goes up one side of the drone and comes back down the other. I definetly did not want it to even think about this being an option.

**Comment by jjshoots** [comment]
Hahahaha that is definitely a problem. I'm not sure what a clean way to fix that is. Termination feels weird while truncation is incorrectly using the truncation term.

**Comment by jjshoots** [comment]
@defrag-bambino an update:

Finally found an extremely stable reward function configuration, and managed to repro the weird collision bug. I'll work out how to solve that now.

https://github.com/user-attachments/assets/a3334cb9-fe8b-42b6-8403-c0bfb6c2c516

**Comment by defrag-bambino** [comment]
> @defrag-bambino an update:
> 
> Finally found an extremely stable reward function configuration, and managed to repro the weird collision bug. I'll work out how to solve that now.
> 
>  screencast.mp4

cool!

File: docs_source/documentation/gym_envs/quadx_ball_in_cup_env.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+```{figure} https://raw.githubusercontent.com/jjshoots/PyFlyt/master/readme_assets/quadx_pole_balance.gif
+    :width: 50%
+```
+
+## Task Description
+
+The goal is to swing up a suspended ball onto the drone, and then bring it to the starting position.
+
+This environment was introduced and contributed by GitHub user [defrag-bambino](https://github.com/jjshoots/PyFlyt/pull/58).
```
Comment (jjshoots) [inline docs_source/documentation/gym_envs/quadx_ball_in_cup_env.md:11]: @defrag-bambino could you check if this affiliation is ok?

**Comment by defrag-bambino** [comment]
I've tried to train it using my stack. It converges to some behavior where it flies up straight and then comes back down and turns itself upside down. Somehow this generates reward of average around 60 - with average episode length of also 60.

https://github.com/user-attachments/assets/bbe2a752-393a-4a08-bad5-4ccfaf75d1f6

**Comment by jet-sony** [comment]
Interesting... Is this a model trained to convergence? On my stack, it learns the optimal behaviour, but doesn't quite get the ball to the center yet because there's no cup - it's on the way.

File: PyFlyt/gym_envs/quadx_envs/quadx_ball_in_cup_env.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+            axis=0,
+        )
+
+    def compute_term_trunc_reward(self) -> None:
+        """Computes the termination, trunction, and reward of the current timestep."""
+        super().compute_base_term_trunc_reward()
+
+        # compute some parameters of the ball
+        # lin_pos: [3,], height: [1,], abs_dist: [1,]
+        ball_rel_lin_pos = self.ball_lin_pos - self.env.state(0)[-1]
```
[the remark this answers is not in the document]
Comment (defrag-bambino) [reply PyFlyt/gym_envs/quadx_envs/quadx_ball_in_cup_env.py:263]: I'm using SheepRL's DreamerV3(_S) with its default settings. Training for maybe 100k steps should be more than enough. Here's my Wrapper for the env.
[sheepRL_env_wrapper.zip](https://github.com/user-attachments/files/16789335/sheepRL_env_wrapper.zip)

[the remark this answers is not in the document]
Comment (jjshoots) [reply PyFlyt/gym_envs/quadx_envs/quadx_ball_in_cup_env.py:263]: I really think it's not trained to completion yet. Despite how powerful Dreamer V3 is, 100k transitions for an environment with sequential task dependencies (hover, swing up, catch, bring to center) generally requires more transitions. This is what the latest development looks like:

[Screencast from 29-08-24 20:47:11.webm](https://github.com/user-attachments/assets/4c9eb92f-117d-4205-b8ea-8b492e6fae9f)

**Pull request merged**